instruments/tests/test_tektronix/test_tekdpo70000.py

```diff
@@ -0,0 +1,2022 @@
+                "DAT:SYNCSOU?"
+            ],
+            [
+                f"{value_io}"
+            ]
+    ) as inst:
+        inst.data_sync_sources = value
+        assert inst.data_sync_sources == value
+
+
+valid_channel_range = [it for it in range(4)]
+
+
+@pytest.mark.parametrize("no", valid_channel_range)
+def test_data_source_channel(no):
+    """Get / set channel as data source."""
+    channel_name = f"CH{no + 1}"
+    with expected_protocol(
+            ik.tektronix.TekDPO70000,
+            [
+                f"DAT:SOU {channel_name}",
+                f"DAT:SOU?"
+            ],
+            [
+                channel_name
+            ]
+    ) as inst:
+        channel = inst.channel[no]
+        inst.data_source = channel
+        assert inst.data_source == channel
+
+
+valid_math_range = [it for it in range(4)]
+
+
+@pytest.mark.parametrize("no", valid_math_range)
+def test_data_source_math(no):
+    """Get / set math as data source."""
+    math_name = f"MATH{no + 1}"
+    with expected_protocol(
+            ik.tektronix.TekDPO70000,
+            [
+                f"DAT:SOU {math_name}",
+                f"DAT:SOU?"
+            ],
+            [
+                math_name
+            ]
+    ) as inst:
+        math = inst.math[no]
+        inst.data_source = math
+        assert inst.data_source == math
+
+
+def test_data_source_ref_not_implemented_error():
+    """Get / set a reference channel raises a NotImplemented error."""
+    ref_name = "REF1"  # example, range not important
+    with expected_protocol(
+            ik.tektronix.TekDPO70000,
+            [
+                f"DAT:SOU?"
+            ],
+            [
+                ref_name
+            ]
+    ) as inst:
+        # getter
+        with pytest.raises(NotImplementedError):
+            print(inst.data_source)
+        # setter
+        with pytest.raises(NotImplementedError):
+            inst.data_source = inst.ref[0]
+
+
+def test_data_source_not_implemented_error():
+    """Get a data source that is currently not implemented."""
+    ds_name = "HHG29"  # example, range not important
+    with expected_protocol(
+            ik.tektronix.TekDPO70000,
+            [
+                f"DAT:SOU?"
+            ],
+            [
+                ds_name
+            ]
+    ) as inst:
+        with pytest.raises(NotImplementedError):
+            print(inst.data_source)
+
+
+def test_data_source_invalid_type():
+    """Raise TypeError when a wrong type is set for data source."""
+    invalid_data_source = 42
+    with expected_protocol(
+            ik.tektronix.TekDPO70000,
+            [
+            ],
+            [
+            ]
+    ) as inst:
+        with pytest.raises(TypeError) as exc_info:
+            inst.data_source = invalid_data_source
+        exc_msg = exc_info.value.args[0]
+        assert exc_msg == f"{type(invalid_data_source)} is not a valid data " \
+                          f"source."
+
+
+def test_data_source_delay():
```

Comment:
This is a funky test, mostly to test that that `time.sleep(0.02)` is really incorporated. A better idea on how to ensure this would be very welcomed, since I don't think it is actually all that meaningful as it is right now.

Reply:
What we can do is remove the `if self._testing` and mock out `time.sleep` for this unit test. Then we just have to assert that it was called.

Reply:
Done, I like that method way better. This is now incorporated into the main test for this routine, the `_delay` test is removed.